include/power_grid_model/math_solver/iterative_current_pf_solver.hpp

```diff
@@ -0,0 +1,306 @@
+            sub_timer = Timer(calculation_info, 2224, "Iterate unknown");
+            // Calculate maximum deviation from previous iteration
+            max_dev = iterate_unknown(output.u);
+            sub_timer.stop();
+        }
+
+        // Invalidate prefactorization when y bus data changes
+        // Invalidate after every calculation now, change later
+        bsr_solver_.invalidate_prefactorization();
+
+        // calculate math result
+        sub_timer = Timer(calculation_info, 2225, "Calculate Math Result");
+        calculate_result(y_bus, input, output);
+
+        // Manually stop timers to avoid "Max number of iterations" to be included in the timing.
+        sub_timer.stop();
+        main_timer.stop();
+
+        const auto key = Timer::make_key(2226, "Max number of iterations");
+        calculation_info[key] = std::max(calculation_info[key], (double)num_iter);
+
+        return output;
+    }
+
+   private:
+    Idx n_bus_;
+    // shared topo data
+    std::shared_ptr<DoubleVector const> phase_shift_;
+    std::shared_ptr<IdxVector const> load_gen_bus_indptr_;
+    std::shared_ptr<IdxVector const> source_bus_indptr_;
+    std::shared_ptr<std::vector<LoadGenType> const> load_gen_type_;
+    //std::vector < ComplexPower < sym>> updated_u_;
+    ComplexValueVector<sym> updated_u_;
+    ComplexValueVector<sym> rhs_;
+    ComplexTensorVector<sym> mat_data_;
+    BSRSolver<DoubleComplex> bsr_solver_;
+
+    void calculate_injected_current(YBus<sym> const& y_bus, PowerFlowInput<sym> const& input,
+                                    ComplexValueVector<sym> const& u) {
+        IdxVector const& load_gen_bus_indptr = *load_gen_bus_indptr_;
+        IdxVector const& source_bus_indptr = *source_bus_indptr_;
+        std::vector<LoadGenType> const& load_gen_type = *load_gen_type_;
+        //ComplexTensorVector<sym> const& ydata = y_bus.admittance();
+        IdxVector const& bus_entry = y_bus.bus_entry();
+
+        // loop individual load/source
+        for (Idx bus_number = 0; bus_number != n_bus_; ++bus_number) {
+            Idx const data_sequence = bus_entry[bus_number];    // Do something for sequence
+            // loop load
+            for (Idx load_number = load_gen_bus_indptr[bus_number]; load_number != load_gen_bus_indptr[bus_number + 1]; ++load_number) {
+                // load type
+                LoadGenType const type = load_gen_type[load_number];
+                switch (type) {
+                    case LoadGenType::const_pq:
+                        // I_inj = conj(S_inj/U) for constant PQ type
+                        rhs_[bus_number] += conj(input.s_injection[load_number] / u[bus_number]);
+                        break;
+                    case LoadGenType::const_y:
+                        // I_inj = conj((S_inj * abs(U*U)) / U) for const impedance type
+                        rhs_[bus_number] += conj(input.s_injection[load_number] * cabs(u[bus_number] * u[bus_number]) / u[bus_number]);
```

Comment:
This is not correct.

 I_inj = conj((S_inj * abs(U)^2) / U) = conj(S_inj) * U

Reply:
Thanks! Fixed that.